packages/eslint-plugin/src/rules/return-await.ts

```diff
@@ -68,64 +72,88 @@ export default createRule({
+      if (tryAncestorResult == null) {
+        return false;
       }
 
-      return false;
-    }
-
-    function inCatch(node: ts.Node): boolean {
-      let ancestor = node.parent as ts.Node | undefined;
+      const { tryStatement, block } = tryAncestorResult;
 
-      while (ancestor && !ts.isFunctionLike(ancestor)) {
-        if (ts.isCatchClause(ancestor)) {
+      switch (block) {
+        case 'try':
+          // Try blocks are always followed by either a catch or finally,
+          // so exceptions thrown here always affect control flow.
           return true;
-        }
-
-        ancestor = ancestor.parent;
-      }
-
-      return false;
-    }
-
-    function isReturnPromiseInFinally(node: ts.Node): boolean {
-      let ancestor = node.parent as ts.Node | undefined;
+        case 'catch':
+          // Exceptions thrown in catch blocks followed by a finally block affect
+          // control flow.
+          if (tryStatement.finallyBlock != null) {
+            return true;
+          }
 
-      while (ancestor && !ts.isFunctionLike(ancestor)) {
-        if (
-          ts.isTryStatement(ancestor.parent) &&
-          ts.isBlock(ancestor) &&
-          ancestor.parent.end === ancestor.end
-        ) {
-          return true;
+          // Otherwise recurse.
+          return affectsExplicitErrorHandling(tryStatement);
+        case 'finally':
+          return affectsExplicitErrorHandling(tryStatement);
+        default: {
+          const __never: never = block;
+          throw new Error(`Unexpected block type: ${String(__never)}`);
         }
-        ancestor = ancestor.parent;
       }
+    }
 
-      return false;
+    interface FindContainingTryStatementResult {
+      tryStatement: ts.TryStatement;
+      block: 'try' | 'catch' | 'finally';
     }
 
-    function hasFinallyBlock(node: ts.Node): boolean {
+    /**
+     * A try _statement_ is the whole thing that encompasses try block,
+     * catch clause, and finally block. This function finds the nearest
+     * enclosing try statement (if present) for a given node, and reports which
+     * part of the try statement the node is in.
+     */
+    function findContainingTryStatement(
+      node: ts.Node,
+    ): FindContainingTryStatementResult | undefined {
+      let child = node;
       let ancestor = node.parent as ts.Node | undefined;
 
       while (ancestor && !ts.isFunctionLike(ancestor)) {
         if (ts.isTryStatement(ancestor)) {
-          return !!ancestor.finallyBlock;
+          let block: 'try' | 'catch' | 'finally';
+          if (child === ancestor.tryBlock) {
+            block = 'try';
+          } else if (child === ancestor.catchClause) {
+            block = 'catch';
+          } else if (child === ancestor.finallyBlock) {
+            block = 'finally';
+          } else {
+            throw new Error(
+              'Child of a try statement must be a try block, catch clause, or finally block',
+            );
```

Comment:
[Testing] If this case can be hit, it should be unit tested and the rule shouldn't crash. But if it can't be hit then it should be removed.

Reply:
Is there a good pattern for how to handle this sort of thing? 

One of the three conditions _must_ be true, or else we have an invalid AST. But there isn't any way to indicate that with the type system that I can think of. 

We _should_ be able to replace the last branch's check with an unconditional `else`, but that hampers readability and maintainability in multiple ways IMO.

We _could_ fail gracefully and `return undefined` in the unreachable case, but that's semantically incorrect and still untestable as far as I can tell. (or even `block = 'UNKNOWN'`, but that would _still_ be untestable and then, what's the caller supposed to do with it?)

Reply:
Yeah it's not ideal. I'd say using a `!` on the variable name is probably fine, or some similar assertion.

Reply:
We have the nullthrows util to add clear runtime assertions to avoid `!`